src/data/nat/hyperoperation.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2023 Mark Andrew Gerads. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mark Andrew Gerads, Junyan Xu, Alex J. Best
+-/
+import tactic.ring
+
+/-!
+# Hyperoperation sequence
+
+This file defines the Hyperoperation sequence.
+`hyperoperation 0 m k = k + 1`
+`hyperoperation 1 m k = m + k`
+`hyperoperation 2 m k = m * k`
+`hyperoperation 3 m k = m ^ k`
+`hyperoperation (n + 1) m (k + 1) = hyperoperation n m (hyperoperation (n + 1) m k)`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Hyperoperation>
+
+## Tags
+
+hyperoperation
+-/
+
+/--
+Implementation of the hyperoperation sequence
+-/
+def hyperoperation : ℕ → ℕ → ℕ → ℕ
+| 0 _ k := k + 1
+| 1 m 0 := m
+| 2 _ 0 := 0
+| (n + 3) _ 0 := 1
+| (n + 1) m (k + 1) := hyperoperation n m (hyperoperation (n + 1) m k)
+
+-- Basic hyperoperation lemmas
+
+lemma hyperoperation_0ab_b.succ (m k : ℕ) : hyperoperation 0 m k = k + 1 :=
+begin
+  rw hyperoperation,
+end
+
+lemma hyperoperation_1a0_a (m : ℕ) : hyperoperation 1 m 0 = m :=
+begin
+  rw hyperoperation,
+end
+
+lemma hyperoperation_2a0_0 (m : ℕ) : hyperoperation 2 m 0 = 0 :=
+begin
+  rw hyperoperation,
+end
+
+lemma hyperoperation_n3a0_1 (n m : ℕ) : hyperoperation (n + 3) m 0 = 1 :=
+begin
+  rw hyperoperation,
+end
+
+lemma hyperoperation_n1ab1_recurse (n m k : ℕ) :
+  hyperoperation (n + 1) m (k + 1) = hyperoperation n m (hyperoperation (n + 1) m k) :=
+begin
+  cases n,
+  rw hyperoperation,
+  cases n;
+  rw hyperoperation,
+end
+
+-- Interesting hyperoperation lemmas
+
+lemma hyperoperation_1_addition (m k : ℕ) : hyperoperation 1 m k = m + k :=
```

Comment:
A better statement would be
```suggestion
lemma hyperoperation_one : hyperoperation 1 = (+) :=
```
Which you can prove by adding `ext` to the start of your proof

Reply:
I will look into doing this, though it might take a while.

Reply:
Done.